tket/tests/Transformations/test_RedundancyRemoval.cpp

```diff
@@ -0,0 +1,305 @@
+          TestGate{OpType::XXPhase3, -0.5, {0, 2, 1}}, true},
+      TestCase{
+          "Cancelling CnXs n=4 (same port order)",
+          TestGate{OpType::CnX, {0, 1, 2, 3, 4}},
+          TestGate{OpType::CnX, {0, 1, 2, 3, 4}}, true},
+      TestCase{
+          "Cancelling CnXs n=4 (swapped control ports)",
+          TestGate{OpType::CnX, {0, 2, 1, 3, 4}},
+          TestGate{OpType::CnX, {2, 3, 0, 1, 4}}, true},
+      TestCase{
+          "Cancelling CnXs n=4 (swapped X port)",
+          TestGate{OpType::CnX, {0, 2, 1, 3, 4}},
+          TestGate{OpType::CnX, {2, 4, 0, 1, 3}}, false},
+      TestCase{
+          "Cancelling CnYs n=4 (same port order)",
+          TestGate{OpType::CnY, {0, 1, 2, 3, 4}},
+          TestGate{OpType::CnY, {0, 1, 2, 3, 4}}, true},
+      TestCase{
+          "Cancelling CnYs n=4 (swapped control ports)",
+          TestGate{OpType::CnY, {0, 2, 1, 3, 4}},
+          TestGate{OpType::CnY, {3, 2, 0, 1, 4}}, true},
+      TestCase{
+          "Cancelling CnYs n=4 (swapped Y port)",
+          TestGate{OpType::CnY, {0, 2, 1, 3, 4}},
+          TestGate{OpType::CnY, {3, 2, 4, 1, 0}}, false},
+      TestCase{
+          "Cancelling CnZs n=4 (same port order)",
+          TestGate{OpType::CnZ, {0, 1, 2, 3, 4}},
+          TestGate{OpType::CnZ, {0, 1, 2, 3, 4}}, true},
+      TestCase{
+          "Cancelling CnZs n=4 (swapped ports, also Z port)",
+          TestGate{OpType::CnZ, {0, 2, 1, 3, 4}},
+          TestGate{OpType::CnZ, {2, 4, 0, 1, 3}}, true},
+      TestCase{
+          "Cancelling PhaseGadgets (same port order)",
+          TestGate{OpType::PhaseGadget, 0.5, {0, 1, 2, 3, 4}},
+          TestGate{OpType::PhaseGadget, -0.5, {0, 1, 2, 3, 4}}, true},
+      TestCase{
+          "Cancelling PhaseGadgets (ports permuted)",
+          TestGate{OpType::PhaseGadget, 0.5, {0, 1, 2, 3, 4}},
+          TestGate{OpType::PhaseGadget, -0.5, {2, 3, 4, 1, 0}}, true},
+      TestCase{
+          "Cancelling CU1s (same port order)",
+          TestGate{OpType::CU1, 0.5, {0, 1}},
+          TestGate{OpType::CU1, -0.5, {0, 1}}, true},
+      TestCase{
+          "Cancelling CU1s (swapped port order)",
+          TestGate{OpType::CU1, 0.5, {0, 1}},
+          TestGate{OpType::CU1, -0.5, {1, 0}}, true});
+  const auto& test_name = test_case.name;
+  const auto& gate1 = test_case.gate1;
+  const auto& gate2 = test_case.gate2;
+  GIVEN(test_name) {
+    if (gate1.param) {
+      test_circuit.add_op<unsigned>(
+          gate1.opType, gate1.param.value(), gate1.args);
+      test_circuit.add_op<unsigned>(
+          gate2.opType, gate2.param.value(), gate2.args);
+    } else {
+      test_circuit.add_op<unsigned>(gate1.opType, gate1.args);
+      test_circuit.add_op<unsigned>(gate2.opType, gate2.args);
+    }
+    Circuit untransformed_circuit(test_circuit);
+    WHEN("calling Transforms::remove_redundancies on circuit") {
+      Transforms::remove_redundancies().apply(test_circuit);
```

Comment:
Could also test the return value here (should be `true` iff circuit has changed).